// @cdklabs/cdk-hyperledger-fabric-network/lib/user.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HyperledgerFabricUser = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
const cdk = require("aws-cdk-lib");
const iam = require("aws-cdk-lib/aws-iam");
const secretsmanager = require("aws-cdk-lib/aws-secretsmanager");
const constructs = require("constructs");
const identity = require("./identity");
/**
 * Creates custom resources to register and enroll users identities
 * with the CA using the fabric-ca-client SDK
 */
class HyperledgerFabricUser extends constructs.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        // Populate instance variables from input properties
        this.userId = props.userId;
        this.affiliation = props.affilitation;
        // Get the custom resources from Identity Resources class
        const customResourceRole = identity.HyperledgerFabricIdentity.customRole;
        const registerProvider = identity.HyperledgerFabricIdentity.userProvider;
        // Ensure the user affiliation includes the member name
        if (!this.affiliation.startsWith(scope.memberName))
            throw new Error('User affiliation is invalid. Affiliation should start with Member name');
        // The user credentials will be stored in these secrets
        const userPasswordSecret = new secretsmanager.Secret(this, `${this.userId}-Password`);
        this.userPrivateKeySecret = new secretsmanager.Secret(this, `${this.userId}-PrivateKey`);
        this.userSignedCertSecret = new secretsmanager.Secret(this, `${this.userId}-SignedCert`);
        // Add secrets arn to the custom role
        customResourceRole.addToPolicy(new iam.PolicyStatement({
            actions: ['secretsmanager:PutSecretValue'],
            resources: [
                userPasswordSecret.secretArn,
                this.userPrivateKeySecret.secretArn,
                this.userSignedCertSecret.secretArn,
            ],
        }));
        const userProp = {
            userId: this.userId,
            affiliation: this.affiliation,
            passwordArn: userPasswordSecret.secretArn,
            privateKeyArn: this.userPrivateKeySecret.secretArn,
            signedCertArn: this.userSignedCertSecret.secretArn,
        };
        // Custom resource to enroll the admin identity using fabric client SDK
        new cdk.CustomResource(this, `${this.userId}Resource`, { serviceToken: registerProvider.serviceToken, properties: userProp });
    }
}
exports.HyperledgerFabricUser = HyperledgerFabricUser;
_a = JSII_RTTI_SYMBOL_1;
HyperledgerFabricUser[_a] = { fqn: "@cdklabs/cdk-hyperledger-fabric-network.HyperledgerFabricUser", version: "0.8.323" };
