// @cdklabs/cdk-hyperledger-fabric-network/lib/utilities.js
"use strict";
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateString = exports.validateInteger = exports.validateAvailabilityZone = exports.validateRegion = exports.getTlsBucket = exports.ENDING_PORT = exports.STARTING_PORT = exports.SUPPORTED_AVAILABILITY_ZONES = exports.SUPPORTED_REGIONS = void 0;
/**
 * Regions where Managed Blockchain is supported, for details see
 * https://aws.amazon.com/managed-blockchain/pricing/hyperledger/
 */
exports.SUPPORTED_REGIONS = [
    'ap-northeast-1',
    'ap-northeast-2',
    'ap-southeast-1',
    'eu-west-1',
    'eu-west-2',
    'us-east-1',
];
/**
 * Map of supported regions to their availability zones
 */
exports.SUPPORTED_AVAILABILITY_ZONES = {
    'ap-northeast-1': [
        'ap-northeast-1a',
        'ap-northeast-1b',
        'ap-northeast-1c',
    ],
    'ap-northeast-2': [
        'ap-northeast-2a',
        'ap-northeast-2b',
        'ap-northeast-2c',
        'ap-northeast-2d',
    ],
    'ap-southeast-1': [
        'ap-southeast-1a',
        'ap-southeast-1b',
        'ap-southeast-1c',
    ],
    'eu-west-1': [
        'eu-west-1a',
        'eu-west-1b',
        'eu-west-1c',
    ],
    'eu-west-2': [
        'eu-west-2a',
        'eu-west-2b',
        'eu-west-2c',
    ],
    'us-east-1': [
        'us-east-1a',
        'us-east-1b',
        'us-east-1c',
        'us-east-1d',
        'us-east-1e',
        'us-east-1f',
    ],
};
/**
 * Starting port of the Network port range
 */
exports.STARTING_PORT = 30001;
/**
 * Ending port of the Network port range
 */
exports.ENDING_PORT = 30004;
/**
 * Returns the S3 Bucket and key that contains the TLS cert file
 */
function getTlsBucket(region) {
    return { bucketName: `${region}.managedblockchain`, key: 'etc/managedblockchain-tls-chain.pem' };
}
exports.getTlsBucket = getTlsBucket;
/**
 * Throw an error if provided region is not in the supported list
 */
function validateRegion(region) {
    if (!exports.SUPPORTED_REGIONS.includes(region)) {
        const regionList = exports.SUPPORTED_REGIONS.join(', ');
        throw new Error(`Managed Blockchain is only available in the following regions: ${regionList}.`);
    }
}
exports.validateRegion = validateRegion;
/**
 * Throw an error if provided availability zone is not in the supported list
 */
function validateAvailabilityZone(region, availabilityZone) {
    const availabililtyZonesForRegion = exports.SUPPORTED_AVAILABILITY_ZONES[region];
    if (typeof availabilityZone === 'undefined' || !availabililtyZonesForRegion.includes(availabilityZone)) {
        const availabilityZoneList = availabililtyZonesForRegion.join(', ');
        throw new Error(`Managed Blockchain in ${region} is only available in the following availability zones: ${availabilityZoneList}.`);
    }
}
exports.validateAvailabilityZone = validateAvailabilityZone;
/**
 * Throw an error if provided number is not an integer, or not with the given range (inclusive)
 */
function validateInteger(value, min, max) {
    if (!Number.isInteger(value))
        return false;
    if (value < min || value > max)
        return false;
    return true;
}
exports.validateInteger = validateInteger;
/**
 * Throw an error if provided string has length with a given range (inclusive),
 * and optionally matches a provided regular expression pattern
 */
function validateString(value, min, max, regexp) {
    if (value.length < min || value.length > max)
        return false;
    if (typeof regexp !== 'undefined' && !value.match(regexp))
        return false;
    return true;
}
exports.validateString = validateString;
